Transfer value in embryonic actor calls

vyzo opened the pull request:
Currently we have no way to transfer value to embryos (or actors that don't exist yet).
This adds an embryonic state check and turns the call into a send if that's the case.
Also adds a check that we don't transfer value in staticcall context, as this would be a violation of readonly semantics.

Fixes https://github.com/filecoin-project/ref-fvm/issues/1004.

vyzo commented on actors/evm/src/interpreter/instructions/call.rs:
```diff
@@ -157,50 +163,76 @@ pub fn call<BS: Blockstore, RT: Runtime<BS>>(
-                        Method::InvokeContractReadOnly
+            // Special casing for embryo/non-existent actors: we just do a SEND (method 0)
+            // which allows us to transfer funds (and create embryos)
+            let is_embryonic = {
```
I am sure this can be rewritten in some funky way that has no ifs, because crabs don't like them...

Aside from that, should we allow non-existent id addrs or is that a bug?

vyzo replied:
Added some logic to not consider a non-existent ID address embryonic, as this is not predictable in general.

mriise replied:
yeah non-existent id addrs are not embryos, embryos should only exist in the context of f4 addresses + abstract accounts, but this is a hack for now...

vyzo commented:
fixed the tests: we need to make sure we can resolve a delegated address to its original actor id (i fixed it in the mock runtime, but we need to make sure the real runtime does the same).

raulk commented on actors/evm/src/interpreter/instructions/call.rs:
```diff
@@ -157,50 +164,77 @@ pub fn call<BS: Blockstore, RT: Runtime<BS>>(
+                if let Some(cid) = system.rt.get_actor_code_cid(&actor_id) {
+                    system.rt.resolve_builtin_actor_type(&cid) == Some(Type::Embryo)
+                } else {
+                    true
```
I think this would be an internal error, could consider panicking here; if we are here it's because the actor exists, so it surely has a CodeCID. However, this works for now, and I'm happy iterating on `next`.

raulk commented on actors/evm/src/interpreter/instructions/call.rs:
```diff
@@ -157,50 +164,77 @@ pub fn call<BS: Blockstore, RT: Runtime<BS>>(
                     &dst_addr,
-                    Method::InvokeContractReadOnly as u64,
-                    // TODO: support IPLD codecs #758
+                    0,
```
Would use the `METHOD_SEND` constant, but can iterate on `next`.

raulk commented on actors/evm/src/interpreter/instructions/call.rs:
```diff
@@ -157,50 +164,77 @@ pub fn call<BS: Blockstore, RT: Runtime<BS>>(
+                        TokenAmount::from(&value),
+                    ),
+
+                    CallKind::DelegateCall => {
```
Not convinced we've covered the case of DELEGATECALL to an EOA correctly. I think it should be handled on the embryonic block, and possibly should noop (as the callee would execute in our context), but this needs deeper analysis. It's quite an edge case, so I'm happy merging as is. I'll open a TODO issue in ref-fvm.

raulk replied:
Ah, nevermind. `DELEGATECALL` does _not_ take a value, so we're good here!

raulk commented on runtime/src/test_utils.rs:
```diff
@@ -506,6 +508,7 @@ impl<BS: Blockstore> MockRuntime<BS> {
         );
         assert_eq!(source.protocol(), Protocol::ID, "source must use ID address protocol");
         self.delegated_addresses.insert(source, target);
+        self.delegated_addresses_source.insert(target, source);
```
At this stage we are better off using a BiMap, but can revisit later.

Stebalien commented on actors/evm/src/interpreter/instructions/call.rs:
```diff
@@ -157,50 +164,77 @@ pub fn call<BS: Blockstore, RT: Runtime<BS>>(
+            let is_embryonic = if dst_addr.protocol() == AddressProtocol::ID {
+                // sanity check: this shouldn't be an ID address, as you can't predict
+                // what actor is gonna sit there.
+                false
```
You can. Once an embryo exists, it has an ID address.

Basically, we need to:

1. Resolve the address.
2. Lookup the code CID for the address.
3. If it doesn't exist or is an embryo, send with method 0.
4. Otherwise, method 2.

Stebalien replied:
@vyzo https://github.com/filecoin-project/builtin-actors/pull/775
